### DISCUSSION TITLE
Renovate might fail to add reviewers with large numbers of teams

### DISCUSSION THREAD

**Discussion by annatindall** [renovatebot/renovate#23066] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.116.6

### If you're self-hosting Renovate, select which platform you are using.

Azure DevOps (dev.azure.com)

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

Hi!

I'm trying to get renovate to assign a reviewer to the PRs it raises in ADO. renovate.json snippet:
```
"reviewers": ["<display-name-of-user>"],
 "assignAutomerge": true
```

This isn't working in our project, which has over 500 teams. I think this is because [addReviewers()](https://github.com/renovatebot/renovate/blob/fbf9d5965d82e3ae4982936b17256bcd84ffb3e0/lib/modules/platform/azure/index.ts#L898) calls [getUserIds()](https://github.com/renovatebot/renovate/blob/fbf9d5965d82e3ae4982936b17256bcd84ffb3e0/lib/modules/platform/azure/index.ts#L808C62-L808C62) to find the reviewer's id.  Without any arguments, the getTeams() call only seems to return the first 100 teams, and in my case, the reviewer's team isn't in that list, so getUserIds() returns an empty list. (Supported by the fact that I never see [this log](https://github.com/renovatebot/renovate/blob/fbf9d5965d82e3ae4982936b17256bcd84ffb3e0/lib/modules/platform/azure/index.ts#L917))

Repro scenario: I'm not sure I can create a project with 500 teams in the free version of ADO - I'll have a look and see if this is possible. I did create a minimal node.js example, which works in the same way as the renovate code and only returns 100 teams:
```
let authHandler = azApi.getPersonalAccessTokenHandler(token); 
let connection = new azApi.WebApi(orgUrl, authHandler);  

let core = await connection.getCoreApi();
let git = await connection.getGitApi();

const repos = await git.getRepositories();
const repo = repos.filter((c) => c.id === "<repo id>")[0];

const teams = await core.getTeams(repo.project.id);

console.log("Number of teams: ", teams.length); // 100
```

(I have also confirmed I was using a valid display name for the reviewer by adding them to the PR with `az repos pr reviewer add`).

### Relevant debug logs

<details><summary>Logs</summary>

```
2023-06-21T03:35:41.6553698Z  INFO: PR created (repository=<project/repo>, branch=<branch>)
2023-06-21T03:35:41.6554357Z        "pr": <id>,
2023-06-21T03:35:41.6554699Z        "prTitle": "<title>"
2023-06-21T03:35:41.6556125Z DEBUG: addParticipants(pr=<id>) (repository=<project/repo>, branch=<branch>)
2023-06-21T03:35:44.8706214Z DEBUG: Added reviewers (repository=<project/repo>, branch=<branch>)
2023-06-21T03:35:44.8706868Z        "reviewers": ["<display-name-of-user>"]
2023-06-21T03:35:44.8709755Z DEBUG: setPrCache() (repository=<project/repo>, branch=<branch>)
2023-06-21T03:35:44.8710349Z DEBUG: Created Pull Request #<id> (repository=<project/repo>, branch=<branch>)
```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by viceice** [comment]
how can we get more pages of that getTeams call?

**Comment by annatindall** [comment]
I was wondering about https://www.npmjs.com/package/@azure/core-paging?

**Comment by ddovile** [comment, marked as answer]
Hi! This issue is still very relevant. In my case none of the reviewers are added either - company has more than 100 teams, [getTeams](https://github.com/renovatebot/renovate/blob/19198355b226a15710d490ca0535041d8b215043/lib/modules/platform/azure/index.ts#L858) return only first 100 teams, the reviewers do not belong to any of them. [Get Teams](https://learn.microsoft.com/en-us/rest/api/azure/devops/core/teams/get-teams?view=azure-devops-rest-7.1&tabs=HTTP) support _$top_ and _$skip_ parameters, so maybe some paging logic can be implemented to check all the teams in the project, not only the first 100?

**Reply by rarkins** [reply]
Yes, that would be reasonable. PRs welcome for such a change. I have created an issue: https://github.com/renovatebot/renovate/issues/28612
